### ISSUE TITLE
FastMonitoringService does not seem to work properly in a multithreaded process

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#43346]
The `{pre,post}ModuleEvent{Acquire,}` functions of the `FastMonitoringService` assumes there is only one module being run in an edm::stream at a time. With HLT using more threads than edm::streams, the framework is running many modules concurrently per edm::stream. The consequence is that the states of the `microstate_` and `microstateAcqFlag_` are presumably incorrect often.

https://github.com/cms-sw/cmssw/blob/3b943c0dbbdf4494cd66064a5a147301f38af295/EventFilter/Utilities/src/FastMonitoringService.cc#L662-L680

**Comment by makortel** [comment]
assign EventFilter/Utilities

**Comment by smorovic** [comment]
@makortel 
yes, it's well known to us that with module parallelism this is no longer accurate.
It still gives some idea of what is going on, like rough fraction of time being idle or if we spend a lot of time in output modules, but this issue is noticeable especially as the "FwkOvhMod" (post-module as last-set state change in the stream).

To improve this, we'd have to sample per TBB thread. I was thinking of that before, but got stuck on how to enumerate thread IDs, since an identifier is needed to fill per-thread array.  Pthread id could be used, which requires remapping it to a position in array (and that might add some overhead with a small map needed).
Or, instead, how feasible would it be to have a thread local variable for each thread from the pool, ranged 0 to num_threads -1?

**Comment by VinInn** [comment]
what about 
tbb::task_arena::current_thread_index()
?
https://spec.oneapi.io/versions/1.0-rev-3/elements/oneTBB/source/task_scheduler/task_arena/this_task_arena_ns.html

**Comment by smorovic** [comment]
It is actually: tbb::this_task_arena::current_thread_index()
I'll try this, thanks.

**Comment by makortel** [comment]
> It is actually: tbb::this_task_arena::current_thread_index()

Indexing by some thread index might work for `microstate_`, but not for `microstateAcqFlag_`. The `*ModuleEventAcquire` and `*ModuleEvent` signals can easily occur in different threads for the same module. In addition, several modules ran in the same thread (and/or stream) can be in the state "`acquire()` finished, but `produce()` has not been called yet".

> Or, instead, how feasible would it be to have a thread local variable for each thread from the pool, ranged 0 to num_threads -1?

This technique has been used earlier elsewhere to index threads. Technically it works, but it makes it very apparent that TBB _does_ remove and add threads to the thread pool (I vaguely recall one offline production job, that failed for other reasons, using > 200 unique threads).

**Comment by smorovic** [comment]
> > It is actually: tbb::this_task_arena::current_thread_index()
> 
> Indexing by some thread index might work for `microstate_`, but not for `microstateAcqFlag_`. The `*ModuleEventAcquire` and `*ModuleEvent` signals can easily occur in different threads for the same module. In addition, several modules ran in the same thread (and/or stream) can be in the state "`acquire()` finished, but `produce()` has not been called yet".
> 

Yes, wasn't sure if that was correct for streams (I suppose modules from the same stream can be scheduled when other module is waiting for results after acquire). In monitoring currently we just sum together those two anyway.
Maybe it is better to repurpose microstateAcqFlag_ to denote separate micro-states for acquire and produce. It would be set to 1 in `preModuleEventAcquire` and set to 0 in `postModuleEventAcquire` (assuming this sequence runs without any yielding and waiting). Also I'm assuming `pre/postModuleEvent` encompasses only `produce()`.

> > Or, instead, how feasible would it be to have a thread local variable for each thread from the pool, ranged 0 to num_threads -1?
> 
> This technique has been used earlier elsewhere to index threads. Technically it works, but it makes it very apparent that TBB _does_ remove and add threads to the thread pool (I vaguely recall one offline production job, that failed for other reasons, using > 200 unique threads).

So that is not the option, but maybe tbb provided index can work.

I implemented that change in this draft PR. Unit test seems to work, but will also need to test it in the proper test system in the coming days:
https://github.com/cms-sw/cmssw/pull/43368

I had to still keep some variables per stream, like processed event accounting per LS. This needs tracking, for each stream, which lumisection it's processing, so it will not work with accounting threads.

Part of the monitoring also tries to do stream-lumisection accounting for microstate (and it will not work correctly with thread tracking), but this is only saved in "slow" json output which we don't use anywhere and that can probably be removed or simplified in subsequent PRs. "Fast" (CSV) mode, which is being used, just samples whatever is currently ongoing, irrespective of LS, and we use this in monitoring;

**Comment by makortel** [comment]
> (assuming this sequence runs without any yielding and waiting). 

Correct, the `preModuleEventAcquire` and `postModuleEventAcquire` are called in the same thread.

> Also I'm assuming `pre/postModuleEvent` encompasses only `produce()`.

Correct. (well, strictly speaking also `filter()` and `analyze()`, but I assume you meant all the "event processing function" flavors)

**Comment by smorovic** [comment]
@makortel
Thanks for details.
I implemented (and tested) thread microstate using `current_thread_index()` as suggested by @VinInn in
https://github.com/cms-sw/cmssw/pull/43368

And I also added thread activity tracking with the task scheduler observer:
https://spec.oneapi.io/versions/latest/elements/oneTBB/source/task_scheduler/task_arena/task_scheduler_observer_cls.html
which is used to modify state (Idle -> Fwk if thread is running something else out of the pre/post module scope, with exception of input source waiting for data which is now a special kind of idle state).

For now per-stream microstate will be kept, I want to compare it with threaded approach in real conditions (data-taking).

**Comment by makortel** [comment]
@smorovic I guess we can close this issue now?
